### PR TITLE
feat: filter cols by lookbehind combination filter

### PR DESCRIPTION
**Pull request by bokajgd**

### CODE REVIEW

File: psycop/common/model_training_v2/trainer/preprocessing/steps/filters.py [hunk 1]
```diff
@@ -16,3 +19,58 @@ def __init__(self, min_age: int, max_age: int, age_col_name: str):
...
+    ):
+        self.lookbehind_combination = lookbehind_combination
+        self.pred_col_prefix = pred_col_prefix
+        self.logger = logger
+
+    def apply(self, input_df: PolarsFrame_T0) -> PolarsFrame_T0:
+        pred_cols_with_lookbehind = [
+            col
+            for col in input_df.columns
+            if col.startswith("pred_") and "within" in col
```
Comment (MartinBernstorff) [inline psycop/common/model_training_v2/trainer/preprocessing/steps/filters.py:40]:
```suggestion
            if col.startswith(self.pred_col_prefix) and "within" in col
```

File: psycop/common/model_training_v2/trainer/preprocessing/steps/filters.py [hunk 2]
```diff
@@ -16,3 +19,58 @@ def __init__(self, min_age: int, max_age: int, age_col_name: str):
...
+        ]
+
+        lookbehinds_in_dataset = {
+            int(re.findall(pattern=r"within_(\d+)_days", string=col)[0])
```
Comment (MartinBernstorff) [inline psycop/common/model_training_v2/trainer/preprocessing/steps/filters.py:44]: We probably want to move the pattern into the __init__ as well. Makes the filter easier to understand within looking into it :-)

File: psycop/common/model_training_v2/trainer/preprocessing/steps/filters.py [hunk 3]
```diff
@@ -16,3 +19,58 @@ def __init__(self, min_age: int, max_age: int, age_col_name: str):
 
     def apply(self, input_df: PolarsFrame_T0) -> PolarsFrame_T0:
         return input_df.filter((self.age >= self.min_age) & (self.age <= self.max_age))
+
+
+@BaselineRegistry.preprocessing.register("lookbehind_combination_filter")
+class LookbehindCombinationFilter(PresplitStep):
```
Comment (MartinBernstorff) [inline psycop/common/model_training_v2/trainer/preprocessing/steps/filters.py:25]:
```suggestion
class LookbehindFilter(PresplitStep):
```

File: psycop/common/model_training_v2/trainer/preprocessing/steps/filters.py [hunk 4]
```diff
@@ -16,3 +19,58 @@ def __init__(self, min_age: int, max_age: int, age_col_name: str):
...
+    def __init__(
+        self,
+        lookbehind_combination: set[int],
```
Comment (MartinBernstorff) [inline psycop/common/model_training_v2/trainer/preprocessing/steps/filters.py:28]:
```suggestion
        lookbehinds: set[int],
```

File: psycop/common/model_training_v2/trainer/preprocessing/steps/filters.py [hunk 5]
```diff
@@ -16,3 +19,58 @@ def __init__(self, min_age: int, max_age: int, age_col_name: str):
...
+        pred_col_prefix: str,
+        logger: BaselineLogger,
+    ):
+        self.lookbehind_combination = lookbehind_combination
```
Comment (MartinBernstorff) [inline psycop/common/model_training_v2/trainer/preprocessing/steps/filters.py:32]:
```suggestion
        self.lookbehinds = lookbehinds
```

File: psycop/common/model_training_v2/trainer/preprocessing/steps/filters.py (hunk 4 above)
Comment (MartinBernstorff) [reply psycop/common/model_training_v2/trainer/preprocessing/steps/filters.py:28]: For brevity :-)

File: psycop/common/model_training_v2/trainer/preprocessing/steps/filters.py [hunk 6]
```diff
@@ -16,3 +19,58 @@ def __init__(self, min_age: int, max_age: int, age_col_name: str):
...
+            )
+
+            lookbehinds_to_keep = self.lookbehind_combination.intersection(
+                lookbehinds_in_dataset,
+            )
+
+            if not lookbehinds_to_keep:
+                self.logger.fail("No predictors left after dropping lookbehinds.")
+                raise ValueError(
+                    "Endng training because no predictors left after dropping lookbehinds.",
```
Comment (MartinBernstorff) [inline psycop/common/model_training_v2/trainer/preprocessing/steps/filters.py:63]:
```suggestion
                    "No predictors left after dropping lookbehinds.",
```

File: psycop/common/model_training_v2/trainer/preprocessing/steps/filters.py [hunk 7]
```diff
@@ -16,3 +19,58 @@ def __init__(self, min_age: int, max_age: int, age_col_name: str):
...
+                )
+
+            self.logger.warn(f"Training on {lookbehinds_to_keep}.")
+        else:
+            lookbehinds_to_keep = self.lookbehind_combination
```
Comment (MartinBernstorff) [inline psycop/common/model_training_v2/trainer/preprocessing/steps/filters.py:68]:
```suggestion
        lookbehinds_to_keep = self.lookbehind_combination
```

Comment (MartinBernstorff) [reply psycop/common/model_training_v2/trainer/preprocessing/steps/filters.py:68]: Don't need the else when the if-branch raises a ValueError :-)

File: psycop/common/model_training_v2/trainer/preprocessing/steps/filters.py [hunk 8]
```diff
@@ -16,3 +19,58 @@ def __init__(self, min_age: int, max_age: int, age_col_name: str):
...
+
+        cols_to_drop = [
+            c
+            for c in pred_cols_with_lookbehind
+            if all(str(l_beh) not in c for l_beh in lookbehinds_to_keep)
```
Comment (MartinBernstorff) [inline psycop/common/model_training_v2/trainer/preprocessing/steps/filters.py:73]: The all(str(l_beh)... was a bit hard for me to parse, so perhaps extract it into its own private method?

File: psycop/common/model_training_v2/trainer/preprocessing/steps/test_v2_filters.py [hunk 9]
```diff
@@ -17,13 +19,32 @@
...
 def test_age_filter(min_age: int, max_age: int, n_remaining: int):
     df = str_to_pl_df(
-        """age,
-    1,
-    2,
-    3,
+        """age,pred_age,
+    1, 2,
+    2, 3,
+    3, 4,
```
Comment (MartinBernstorff) [inline psycop/common/model_training_v2/trainer/preprocessing/steps/test_v2_filters.py:25]: Why this change?

File: psycop/common/model_training_v2/trainer/preprocessing/steps/test_v2_filters.py [hunk 10]
```diff
@@ -17,13 +19,32 @@
...
+    )
+
+    logger = TerminalLogger()
+    result = LookbehindCombinationFilter(
+        lookbehind_combination={2, 3},
+        pred_col_prefix="pred_",
+        logger=logger,
+    ).apply(df)
+
+    assert len(result.columns) == 3
```
Comment (MartinBernstorff) [inline psycop/common/model_training_v2/trainer/preprocessing/steps/test_v2_filters.py:50]: Great test!

File: psycop/common/model_training_v2/trainer/preprocessing/steps/filters.py [hunk 11]
```diff
@@ -16,3 +19,58 @@ def __init__(self, min_age: int, max_age: int, age_col_name: str):
 
     def apply(self, input_df: PolarsFrame_T0) -> PolarsFrame_T0:
         return input_df.filter((self.age >= self.min_age) & (self.age <= self.max_age))
+
+
+@BaselineRegistry.preprocessing.register("lookbehind_combination_filter")
```
Comment (MartinBernstorff) [inline psycop/common/model_training_v2/trainer/preprocessing/steps/filters.py:24]:
```suggestion
@BaselineRegistry.preprocessing.register("lookbehind_filter")
```

**Review by MartinBernstorff — COMMENTED** [review]
Excellent work Jacob! Just a handful of readability-suggestions, the it's ready for merge!

File: psycop/common/model_training_v2/trainer/preprocessing/steps/filters.py [hunk 12]
```diff
@@ -1,6 +1,9 @@
+import re
```
Comment (MartinBernstorff) [inline psycop/common/model_training_v2/trainer/preprocessing/steps/filters.py:1]: Perhaps we already want to split these filters into a row_filters file and a col_filters file 👍

**Review by MartinBernstorff — APPROVED** [review]

**Comment by bokajgd** [comment]
Thanks Martin, as you experienced, I sent out this PR a bit too soon. Didn't really run through the code after writing it, hence the many small 'mistakes'. Anyway, the setup is great to work with and I will happily take on more taksks to write up or adapt desired functionalitid

File: psycop/common/model_training_v2/trainer/preprocessing/steps/filters.py (hunk 3 above)
Comment (bokajgd) [reply psycop/common/model_training_v2/trainer/preprocessing/steps/filters.py:25]: I guess we might need another col filter based on lookbehind, e.g. `LookbehindExceedsLookbehindFilter` so maybe we should stick to the more specific name.

File: psycop/common/model_training_v2/trainer/preprocessing/steps/test_v2_filters.py (hunk 9 above)
Comment (bokajgd) [reply psycop/common/model_training_v2/trainer/preprocessing/steps/test_v2_filters.py:25]: No reason

File: psycop/common/model_training_v2/trainer/preprocessing/steps/filters.py (hunk 7 above)
Comment (bokajgd) [reply psycop/common/model_training_v2/trainer/preprocessing/steps/filters.py:68]: Yes we do, becasue the if-branch that raises the ValueError is a sub-if-branch. The else here relates back to the first if.

**Comment by bokajgd** [comment]
I just have a few responses that you might want to re-respond to? Otherwise it should be ready for merge.

**Comment by bokajgd** [comment]
Ohh, forgot to split the filters into two files. One sec!

File: psycop/common/model_training_v2/trainer/preprocessing/steps/filters.py (hunk 3 above)
Comment (MartinBernstorff) [reply psycop/common/model_training_v2/trainer/preprocessing/steps/filters.py]: Hmm, I see. Perhaps add something with column in the name (both for the registry and for the class). Will make the configs much easier to read in the future :-)

Other than that, ready for merge!

**Comment by MartinBernstorff** [comment]
Great! Will make a tiny change myself, and then merge it :D

Great work!
